File: src/tools/write_file.ts
import { z } from 'zod';
import { writeFile, mkdir, stat } from 'fs/promises';
import { dirname } from 'path';
import type { Tool } from './types';

const WriteFileParams = z.object({
  path: z.string().min(1).describe('Absolute or relative path of the file to write.'),
  content: z.string().describe('Full file contents to write.'),
  overwrite: z
    .boolean()
    .optional()
    .default(false)
    .describe('If true, allow overwriting an existing file. Defaults to false.'),
});

export const writeFileTool: Tool = {
  name: 'write_file',
  description:
    'Create a new file with the given contents. Refuses to overwrite existing files unless `overwrite: true` is passed. ' +
    'Parent directories are created automatically. Use this for new files; for existing files prefer `edit_file`.',
  parameters: WriteFileParams,
  safety: {
    sideEffect: 'write',
    permission: 'prompt',
    reason: 'Creates or overwrites files.',
  },
  async execute(args) {
    const { path, content, overwrite } = WriteFileParams.parse(args);

    // Use stat() for an existence check: a zero-byte file should still be
    // considered existing, otherwise write_file would silently overwrite
    // touched-but-empty files.
    let existed = false;
    try {
      await stat(path);
      existed = true;
      if (!overwrite) {
        return `Error: ${path} already exists. Pass overwrite: true to replace it.`;
      }
    } catch {
      // File does not exist; safe to create.
    }

    try {
      await mkdir(dirname(path), { recursive: true });
      await writeFile(path, content, 'utf-8');
      return existed
        ? `Overwrote ${path} (${content.length} bytes).`
        : `Created ${path} (${content.length} bytes).`;
    } catch (err: any) {
      return `Error writing file ${path}: ${err.message}`;
    }
  },
};

File: internal/tools/write_file.go
package tools

import (
	"context"
	"fmt"
	"os"
	"path/filepath"
)

type writeFileTool struct {
	baseTool
	workspaceRoot string
}

func NewWriteFileTool(workspaceRoot string) Tool {
	return writeFileTool{
		baseTool: baseTool{
			name:        "write_file",
			description: "Create a new file, refusing to overwrite existing files unless overwrite is true.",
			parameters: Schema{
				Type: "object",
				Properties: map[string]PropertySchema{
					"path":      {Type: "string", Description: "Absolute or relative path of the file to write."},
					"content":   {Type: "string", Description: "Full file contents to write."},
					"overwrite": {Type: "boolean", Description: "Whether to allow overwriting an existing file.", Default: false},
				},
				Required:             []string{"path", "content"},
				AdditionalProperties: false,
			},
			safety: promptSafety(SideEffectWrite, "Creates or overwrites files."),
		},
		workspaceRoot: normalizeWorkspaceRoot(workspaceRoot),
	}
}

func (tool writeFileTool) Run(_ context.Context, args map[string]any) Result {
	requestedPath, err := aliasedStringArg(args, []string{"path", "file", "file_path", "filename"}, "", true, false)
	if err != nil {
		return errorResult("Error: Invalid arguments for write_file: " + err.Error())
	}
	content, err := fileContentArg(args)
	if err != nil {
		return errorResult("Error: Invalid arguments for write_file: " + err.Error())
	}
	overwrite, err := boolArg(args, "overwrite", false)
	if err != nil {
		return errorResult("Error: Invalid arguments for write_file: " + err.Error())
	}

	absolutePath, relativePath, err := resolveWorkspaceTargetPath(tool.workspaceRoot, requestedPath)
	if err != nil {
		return errorResult("Error writing file " + requestedPath + ": " + err.Error())
	}

	existed := false
	if _, err := os.Stat(absolutePath); err == nil {
		existed = true
		if !overwrite {
			return errorResult("Error: " + relativePath + " already exists. Pass overwrite: true to replace it.")
		}
	} else if !os.IsNotExist(err) {
		return errorResult("Error writing file " + relativePath + ": " + err.Error())
	}

	if err := os.MkdirAll(filepath.Dir(absolutePath), 0o755); err != nil {
		return errorResult("Error writing file " + relativePath + ": " + err.Error())
	}
	if err := recheckWorkspaceWriteTarget(tool.workspaceRoot, requestedPath); err != nil {
		return errorResult("Error writing file " + relativePath + ": " + err.Error())
	}
	if err := os.WriteFile(absolutePath, []byte(content), 0o644); err != nil {
		return errorResult("Error writing file " + relativePath + ": " + err.Error())
	}

	verb := "Created"
	if existed {
		verb = "Overwrote"
	}
	summary := fmt.Sprintf("%s %s (%d bytes).", verb, relativePath, len([]byte(content)))
	result := okResult(summary)
	result.ChangedFiles = []string{relativePath}
	result.Display = Display{Summary: summary, Kind: "file"}
	return result
}

// fileContentArg reads the file body from "content" or a common alias that weaker
// models sometimes use instead (contents/text/body/data/file_content). It
// delegates to the shared aliasedStringArg so the present-but-non-string type
// error ("content must be a string") and the required-but-missing error
// ("content is required") stay consistent with every other tool. An empty string
// is allowed (writing an empty file), so allowEmpty is true.
func fileContentArg(args map[string]any) (string, error) {
	return aliasedStringArg(args, []string{"content", "contents", "text", "body", "data", "file_content"}, "", true, true)
}
